Implement a general statistics endpoint for Community monitoring

denarch22 opened the pull request:
## 🛠️ Changes

- Add GET endpoint /GeneralStatistics with 3 main metrics:
  * Total active initiatives
  * Total people involved in active initiatives
  * Total area under monitoring in hectares
  
- Implement GeneralStatisticsService with business logic
- Add GeneralStatisticsDto for response structure
- Create LocationAreaService for geographic area calculations
- Implement GeometryUtils for area conversions
- Add migration for polygon_area field in Initiative
- Include complete unit and integration tests
- Swagger documentation with response examples
- Fix StyleCop and Code Analysis warnings

## 📝 Associated issues
resolve [LIB-323](https://linear.app/biodev/issue/LIB-323/monitoreo-comunitario-endpoint-de-cifras-generales)

denarch22 committed:
feat: Implementar endpoint de estadísticas generales para monitoreo comunitario

- Agregar endpoint GET /GeneralStatistics con 3 métricas principales:
  * Total de iniciativas activas
  * Total de personas involucradas en iniciativas activas
  * Área total bajo monitoreo en hectáreas

- Implementar GeneralStatisticsService con lógica de negocio
- Agregar GeneralStatisticsDto para estructura de respuesta
- Crear LocationAreaService para cálculos de área geográfica
- Implementar GeometryUtils para conversiones de área
- Agregar migración para campo polygon_area en Initiative
- Incluir tests unitarios e integración completos
- Documentación Swagger con ejemplos de respuesta
- Corregir warnings de StyleCop y Code Analysis

Closes LIB-323

denarch22 committed:
Remove test projects and update initiative polygon handling

Deleted Application.Tests and WebApi.Tests projects and related test files. Updated InitiativeController to accept polygon updates as object and serialize to GeoJSON, changed content type to application/json. Improved logging format and exception handling in InitiativeService. Updated migration files to consolidate polygon area changes.

cgalvist commented on src/Application/DTOs/Reports/GeneralStatisticsDto.cs:
```diff
@@ -0,0 +1,22 @@
+public class GeneralStatisticsDto
+{
+    /// <summary>
+    /// Número total de iniciativas activas en la plataforma.
```
Por favor, cambiar el comentario a Inglés

cgalvist commented on src/Application/DTOs/Reports/GeneralStatisticsDto.cs:
```diff
@@ -0,0 +1,22 @@
+    public int TotalActiveInitiatives { get; set; }
+
+    /// <summary>
+    /// Número de personas involucradas en iniciativas activas.
```
Por favor, cambiar el comentario a Inglés

cgalvist commented on src/Application/DTOs/Reports/GeneralStatisticsDto.cs:
```diff
@@ -0,0 +1,22 @@
+    public int TotalPeopleInvolved { get; set; }
+
+    /// <summary>
+    /// Área total de las iniciativas activas en hectáreas.
```
Por favor, cambiar el comentario a Inglés

cgalvist commented on src/Infrastructure/Persistence/Config/Entities/Initiatives/InitiativeConfig.cs:
```diff
@@ -56,6 +56,11 @@ public void Configure(EntityTypeBuilder<Initiative> builder)
             .HasColumnType("geometry(Point, 4326)")
             .IsRequired();
 
+        builder.Property(i => i.PolygonArea)
```
- Aquí considero que podemos aprovechar para colocar el valor por defecto `0`, ya que `PolygonArea` es un campo obligatorio.
- ¿es necesario configurar el tipo de columna `decimal(15,6)`? Me parece que definiendo el tipo `decimal` en C# las migraciones cambian el tipo de variable, y usar la función `HasColumnType` no sería necesario en este caso

cgalvist commented on src/Application/Application.csproj:
```diff
@@ -11,6 +11,7 @@
   <ItemGroup>
     <PackageReference Include="FluentValidation" Version="12.0.0" />
     <PackageReference Include="Microsoft.AspNetCore.OData" Version="8.3.1" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="8.0.0" />
```
En la arquitectura por capas que estamos manejando, `EntityFramework` no debería utilizarse en el proyecto `Application`. Solo se está usando en el proyecto `Infrastructure`.

Para consultas SQL personalizadas, puedes usar especificaciones con Ardalis o crear métodos nuevos en el repositorio `InitiativeRepository`.

Si tienes alguna duda me puedes contactar por interno y lo revisamos.

cgalvist commented on src/Application/Services/GeneralStatisticsService.cs:
```diff
@@ -0,0 +1,84 @@
+            // Get active initiatives query
+            var activeInitiativesQuery = initiativeRepository.GetQueryable().Where(i => i.Enabled);
+
+            // 1. Número total de iniciativas activas en la plataforma
```
- Por favor, cambiar el comentario a Inglés
- Ya que esta función realiza muchos cálculos diferentes, recomiendo separar cada cálculo en una función privada

cgalvist commented on src/Application/Services/GeneralStatisticsService.cs:
```diff
@@ -0,0 +1,84 @@
+            var activeInitiatives = await initiativeRepository.QueryToListAsync(activeInitiativesQuery, ct);
+            statistics.TotalActiveInitiatives = activeInitiatives.Count;
+
+            // 2. Número de personas involucradas en iniciativas activas
```
- Por favor, cambiar el comentario a Inglés
- Ya que esta función realiza muchos cálculos diferentes, recomiendo separar cada cálculo en una función privada

cgalvist commented on src/Application/Services/GeneralStatisticsService.cs:
```diff
@@ -0,0 +1,84 @@
+            var peopleInActiveInitiatives = await initiativeUserRepository.QueryToListAsync(peopleInActiveInitiativesQuery, ct);
+            statistics.TotalPeopleInvolved = peopleInActiveInitiatives.Count;
+
+            // 3. Área total de las iniciativas activas en hectáreas
```
- Por favor, cambiar el comentario a Inglés
- Ya que esta función realiza muchos cálculos diferentes, recomiendo separar cada cálculo en una función privada

cgalvist commented on src/Application/Services/GeneralStatisticsService.cs:
```diff
@@ -0,0 +1,84 @@
+            var initiativesWithArea = await initiativeRepository.QueryToListAsync(totalAreaQuery, ct);
+            var totalAreaInSquareKm = initiativesWithArea.Sum(i => i.PolygonArea);
+
+            // Convertir de km² a hectáreas (1 km² = 100 hectáreas)
```
Por favor, cambiar el comentario a Inglés

cgalvist commented on src/Application/Services/GeneralStatisticsService.cs:
```diff
@@ -0,0 +1,84 @@
+                ResponseBody = statistics,
+            };
+        }
+        catch (Exception ex)
```
Aquí considero que no deberíamos usar el try catch como un comodín
¿En qué casos concretos puede fallar esta función?

cgalvist reviewed (changes requested):
Hola.

Por el momento he realizado una revisión a nivel de código, no de funcionalidad.

También te quería preguntar ¿porqué se eliminó el archivo `src/Infrastructure/Persistence/Migrations/20250919140408_AddInitiativeJoinInvitations.cs`?

denarch22 commented:
se realizaron las revisiones y ajustes al código:

✅ Comentarios en Inglés: Todos los comentarios traducidos
✅ Arquitectura Limpia: Entity Framework removido de Application layer
✅ Código Refactorizado: Métodos privados para mejor organización
✅ Validaciones Robustas: Verificación de integridad de datos con if-else
✅ Consistencia: Manejo de errores alineado con el resto del proyecto
✅ Migraciones Restauradas: Archivo eliminado recuperado correctamente
✅ Configuración Optimizada: Valores por defecto y tipos correctos

denarch22 committed:
Refactor general statistics logic and add join invitations migration

Refactored GeneralStatisticsService to use new IInitiativeRepository methods for statistics calculation, improving separation of concerns and code clarity. Added methods to IInitiativeRepository and implemented them in InitiativeRepository. Updated DTO comments to English. Added migration for initiative join invitations and guests. Set default value for PolygonArea in InitiativeConfig.

cgalvist commented on src/Application/Services/GeneralStatisticsService.cs:
```diff
@@ -0,0 +1,108 @@
+﻿namespace IAVH.BioTablero.CM.Application.Services;
+
+using System;
+using System.Linq;
```
Esto no se está usando

cgalvist commented on src/Application/Services/GeneralStatisticsService.cs:
```diff
@@ -0,0 +1,108 @@
+using IAVH.BioTablero.CM.Application.DTOs.Reports;
+using IAVH.BioTablero.CM.Application.Interfaces.Services;
+using IAVH.BioTablero.CM.Application.Utils;
+using IAVH.BioTablero.CM.Core.Domain.Entities.Initiatives;
```
Esto no se está usando

cgalvist commented on src/Application/Services/Statistics/GeneralStatisticsService.cs:
¿También podrías colocar este servicio en una carpeta porfa? Podría ser `Statistics`, `Stats` o algún nombre similar

cgalvist commented on src/Application/Interfaces/Services/IGeneralStatisticsService.cs:
```diff
@@ -0,0 +1,20 @@
+using System.Threading;
+using System.Threading.Tasks;
+
+using IAVH.BioTablero.CM.Application.DTOs.Reports;
```
Esto no se está usando

cgalvist commented on src/Application/Interfaces/Services/ILocationAreaService.cs:
Creo que para evitar tener un archivo adicional, sería bueno que los métodos de esta interfaz estén en la interfaz del servicio `Location` ¿Qué opinas?

cgalvist commented on src/Application/Services/Geo/LocationAreaService.cs:
```diff
@@ -0,0 +1,116 @@
+using IAVH.BioTablero.CM.Core.Domain.Entities.Geo;
+using IAVH.BioTablero.CM.Core.Interfaces.Repositories;
+
+using NetTopologySuite.Geometries;
```
Esto no se está usando

cgalvist commented on src/Application/Services/Geo/LocationAreaService.cs:
Creo que para evitar tener un archivo adicional, sería bueno que los métodos de este servicio estén en el servicio de `Location`

cgalvist commented on src/WebApi/Controllers/Rest/Reports/GeneralStatisticsController.cs:
```diff
@@ -0,0 +1,43 @@
+using System.Threading;
+using System.Threading.Tasks;
+
+using IAVH.BioTablero.CM.Application.DTOs.Reports;
```
Esto no se está usando

cgalvist commented on src/WebApi/Controllers/Rest/Reports/GeneralStatisticsController.cs:
```diff
@@ -0,0 +1,43 @@
+using IAVH.BioTablero.CM.WebApi.Config.DocsSetup.Examples.Reports;
+using IAVH.BioTablero.CM.WebApi.Interfaces;
+
+using Microsoft.AspNetCore.Authorization;
```
Esto no se está usando

cgalvist reviewed (changes requested):
Hola @denarch22. Muchas gracias por tus aportes.

He añadido algunos comentarios adicionales. El tema que más me preocupaba era el de las consultas SQL en la capa de aplicación pero ya vi que lo solucionaste 🙂.

Hay otro tema que me preocupa y es que en el archivo `20250919140408_AddInitiativeJoinInvitations.cs` se eliminaron muchas líneas de código y este archivo debería estar intacto, ya que es la migración de un desarrollo anterior a este.

Quedo atento.

denarch22 committed:
Refactor location area logic into LocationService

Removed ILocationAreaService and LocationAreaService, moving area calculation methods into ILocationService and LocationService. Updated InitiativeService to use ILocationService for area calculations. Cleaned up related dependency registrations and interface usages. Also moved GeneralStatisticsService to a new Statistics namespace and updated .editorconfig with unused code detection rules.

denarch22 committed:
Enable XML documentation file generation

Set GenerateDocumentationFile to true in Application, Core, and Infrastructure projects to produce XML documentation files. Added missing XML documentation parameters in JoinRequestService for improved code clarity.

cgalvist commented on src/Application/Services/Geo/LocationService.cs:
```diff
@@ -79,4 +81,82 @@ public async Task<CustomWebResponse> GetPolygonAsync(int id, CancellationToken c
+    /// <param name="locationId">Location ID.</param>
+    /// <param name="ct">Cancellation token.</param>
+    /// <returns>Area in square kilometers.</returns>
+    public async Task<double> CalculateAreaForLocationAsync(int locationId, CancellationToken ct = default)
```
No veo que esta función se esté utilizando

cgalvist commented on src/Application/Services/Geo/LocationService.cs:
```diff
@@ -79,4 +81,82 @@ public async Task<CustomWebResponse> GetPolygonAsync(int id, CancellationToken c
+    /// <param name="locationCodes">Collection of location codes.</param>
+    /// <param name="ct">Cancellation token.</param>
+    /// <returns>Total area in square kilometers.</returns>
+    public async Task<double> CalculateTotalAreaForLocationCodesAsync(IEnumerable<string> locationCodes, CancellationToken ct = default)
```
No veo que esta función se esté utilizando

cgalvist commented on src/Application/Utils/GeometryUtils.cs:
```diff
@@ -0,0 +1,77 @@
+    /// </summary>
+    /// <param name="geometries">Collection of geometries.</param>
+    /// <returns>Total area in square kilometers.</returns>
+    public static double CalculateTotalAreaInSquareKilometers(params Geometry[] geometries)
```
No veo que esta función se esté utilizando

cgalvist commented on src/Application/Application.csproj:
```diff
@@ -2,6 +2,7 @@
 
   <PropertyGroup>
     <RootNamespace>IAVH.BioTablero.CM.Application</RootNamespace>
+    <GenerateDocumentationFile>true</GenerateDocumentationFile>
```
Veo que en `Application`, `Core` e `Infrastructure` marcaste esta opción. Por curiosidad,  ¿para qué lo estás haciendo? 😅

denarch22 replied:
Es una propiedad de MSBuild que:
Genera archivos XML de documentación durante la compilación
Habilita análisis de código más avanzado (como IDE0005)
Permite detectar usings innecesarios en tiempo de compilación

cgalvist replied:
Perfecto.

Si quieres que todos los proyectos tengan esa propiedad, puedes hacerlo editando el archivo `Directory.Build.props` agregando esa propiedad, y eliminándola de todos los archivos `.csproj`.

Va a tener el mismo efecto que ahora, pero no se va a duplicar esa línea 😉

cgalvist commented on .editorconfig:
Agradezco mucho que hayas agregado más reglas para mejorar la calidad del código 😉.

Sin embargo, hay reglas duplicadas:
- IDE0005
- ca1822
- ca1823
- CA1804
- CA1801

denarch22 replied:
Hola Cesar gracias por tus comentarios, ya realice los ajustes indicados.

cgalvist reviewed (changes requested):
Hola.

Probé el proyecto en mi equipo creando toda la base de datos de cero y me funciona.

Sin embargo, te dejo algunas observaciones adicionales.

denarch22 committed:
Remove unused location area calculation methods

Deleted unused methods for calculating area by location ID and codes from ILocationService, LocationService, and GeometryUtils. This streamlines the codebase by removing redundant functionality.

denarch22 committed:
Centralize documentation file generation in build props

Moved the <GenerateDocumentationFile>true</GenerateDocumentationFile> property from individual project files to Directory.Build.props for consistent documentation generation across all projects.
